From 56e94cee4f70ba486b954062d098c65eafdadc6b Mon Sep 17 00:00:00 2001
From: Jesse Chan <jc@lineageos.org>
Date: Sat, 21 Apr 2018 23:54:05 -0700
Subject: [PATCH 3/4] Settings: Add an option to let pre-O apps to use full
 screen aspect ratio

When an app target pre-O releases, the default max aspect ratio
is 1.86:1 which leads to ugly black areas on devices that have
screens with higher aspect ratio (for example Galaxy S8/S9).

This change adds an option to allow users to change default
aspect ratio for pre-O apps to 2.1 which would fit recent devices.

Change-Id: I398d21ffd137faa0be7fb11a138683813f80f3fe
---
 res/values/strings.xml                        |  6 ++
 res/xml/display_settings.xml                  |  5 ++
 src/com/android/settings/DisplaySettings.java |  2 +
 ...ScreenAspectRatioPreferenceController.java | 58 +++++++++++++++++++
 4 files changed, 71 insertions(+)
 create mode 100644 src/com/android/settings/display/FullScreenAspectRatioPreferenceController.java

diff --git a/res/values/strings.xml b/res/values/strings.xml
index 9baaa7c561..0c04c9d4f5 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -9054,4 +9054,10 @@
 
     <!-- Note displayed when certain features are not available on low ram devices. [CHAR LIMIT=NONE] -->
     <string name="disabled_low_ram_device">This feature is not available on this device</string>
+
+    <!-- Label for full screen aspect ratio setting [CHAR LIMIT=30] -->
+    <string name="full_screen_aspect_ratio_title">Full screen aspect ratio</string>
+
+    <!-- Summary for full screen aspect ratio setting [CHAR LIMIT=90] -->
+    <string name="full_screen_aspect_ratio_summary">Use apps in the full screen aspect ratio</string>
 </resources>
diff --git a/res/xml/display_settings.xml b/res/xml/display_settings.xml
index 02645f7d2a..14c623366b 100644
--- a/res/xml/display_settings.xml
+++ b/res/xml/display_settings.xml
@@ -157,6 +157,11 @@
         android:defaultValue="false"
         lineage:requiresFeature="lineagehardware:FEATURE_HIGH_TOUCH_SENSITIVITY" />
 
+    <SwitchPreference
+        android:key="full_screen_aspect_ratio"
+        android:title="@string/full_screen_aspect_ratio_title"
+        android:summary="@string/full_screen_aspect_ratio_summary" />
+
     <!-- Use LineageParts for styles 
     <ListPreference
         android:key="theme"
diff --git a/src/com/android/settings/DisplaySettings.java b/src/com/android/settings/DisplaySettings.java
index 027954c381..721844e97f 100644
--- a/src/com/android/settings/DisplaySettings.java
+++ b/src/com/android/settings/DisplaySettings.java
@@ -29,6 +29,7 @@ import com.android.settings.display.BrightnessLevelPreferenceController;
 import com.android.settings.display.CameraGesturePreferenceController;
 import com.android.settings.display.ColorModePreferenceController;
 import com.android.settings.display.FontSizePreferenceController;
+import com.android.settings.display.FullScreenAspectRatioPreferenceController;
 import com.android.settings.display.LiftToWakePreferenceController;
 import com.android.settings.display.NightDisplayPreferenceController;
 import com.android.settings.display.NightModePreferenceController;
@@ -109,6 +110,7 @@ public class DisplaySettings extends DashboardFragment {
         //controllers.add(new ThemePreferenceController(context));
         controllers.add(new BrightnessLevelPreferenceController(context, lifecycle));
         controllers.add(new ColorModePreferenceController(context));
+        controllers.add(new FullScreenAspectRatioPreferenceController(context));
         return controllers;
     }
 
diff --git a/src/com/android/settings/display/FullScreenAspectRatioPreferenceController.java b/src/com/android/settings/display/FullScreenAspectRatioPreferenceController.java
new file mode 100644
index 0000000000..269c9a9f05
--- /dev/null
+++ b/src/com/android/settings/display/FullScreenAspectRatioPreferenceController.java
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2016 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.android.settings.display;
+
+import android.content.Context;
+import android.provider.Settings;
+import android.support.v14.preference.SwitchPreference;
+import android.support.v7.preference.Preference;
+
+import com.android.settings.core.PreferenceControllerMixin;
+import com.android.settingslib.core.AbstractPreferenceController;
+
+public class FullScreenAspectRatioPreferenceController extends AbstractPreferenceController implements
+        PreferenceControllerMixin, Preference.OnPreferenceChangeListener {
+
+    private static final String KEY_FULL_SCREEN_ASPECT_RATIO = "full_screen_aspect_ratio";
+
+    public FullScreenAspectRatioPreferenceController(Context context) {
+        super(context);
+    }
+
+    @Override
+    public String getPreferenceKey() {
+        return KEY_FULL_SCREEN_ASPECT_RATIO;
+    }
+
+    @Override
+    public boolean isAvailable() {
+        return mContext.getResources().getBoolean(
+                com.android.internal.R.bool.config_haveHigherAspectRatioScreen);
+    }
+
+    @Override
+    public void updateState(Preference preference) {
+        int value = Settings.Secure.getInt(
+                mContext.getContentResolver(), Settings.Secure.FULL_SCREEN_ASPECT_RATIO, 0);
+        ((SwitchPreference) preference).setChecked(value != 0);
+    }
+
+    @Override
+    public boolean onPreferenceChange(Preference preference, Object newValue) {
+        boolean value = (Boolean) newValue;
+        Settings.Secure.putInt(
+                mContext.getContentResolver(), Settings.Secure.FULL_SCREEN_ASPECT_RATIO, value ? 1 : 0);
+        return true;
+    }
+}
-- 
2.17.0

